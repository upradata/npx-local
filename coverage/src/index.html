<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../index.html">All files</a> src
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">36.73% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>18/49</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">16.67% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>2/12</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">18.75% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>3/16</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">36.96% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>17/46</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="node-project.ts"><a href="node-project.ts.html">node-project.ts</a></td>
	<td data-value="19.44" class="pic low"><div class="chart"><div class="cover-fill" style="width: 19%;"></div><div class="cover-empty" style="width:81%;"></div></div></td>
	<td data-value="19.44" class="pct low">19.44%</td>
	<td data-value="36" class="abs low">7/36</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="12" class="abs low">0/12</td>
	<td data-value="20.59" class="pct low">20.59%</td>
	<td data-value="34" class="abs low">7/34</td>
	</tr>

<tr>
	<td class="file high" data-value="read-json5.ts"><a href="read-json5.ts.html">read-json5.ts</a></td>
	<td data-value="84.62" class="pic high"><div class="chart"><div class="cover-fill" style="width: 84%;"></div><div class="cover-empty" style="width:16%;"></div></div></td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="4" class="abs medium">2/4</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="4" class="abs medium">3/4</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="12" class="abs high">10/12</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Nov 21 2019 00:18:47 GMT+0100 (GMT+01:00)
</div>
</div>
<script src="../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../sorter.js"></script>
<script src="../block-navigation.js"></script>
</body>
</html>
