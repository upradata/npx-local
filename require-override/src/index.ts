export * from './require';
export * from './full-path';
export * from './export';
export * from './types';
