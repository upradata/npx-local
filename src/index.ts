#!/usr/bin/env node

import path from 'path';
import type { SettingsPrivate } from './settings.private';


if (process.env.NODE_ENV === 'production') {

    import('./yargs').then(({ runCommand }) => runCommand());

} else {

    const main = async () => {

        // To be independent from self package
        const settings: SettingsPrivate = await import('../settings.private.json');
        const { libraryFolder } = settings.upradata;

        const { RequireOverride } = await import(path.join(libraryFolder, 'require-override'));
        const { lookForLocalPackages } = await import('./local-packages');

        const localPackages = await lookForLocalPackages(libraryFolder);

        new RequireOverride().start({
            module: (requestPath: string) => requestPath.startsWith('@upradata'),
            newModule: (requestPath: string) => {
                const segments = requestPath.split('/');
                const packageName = segments.slice(0, 2).join('/'); // @upradata/name

                const foundPackage = localPackages.find(p => p.package.name === packageName);
                return foundPackage ? foundPackage.folder : undefined;
            }
        });

        const { runCommand } = await import('./yargs');

        runCommand();
    };

    main();

    /* (async function f() {
        const libraryFolder = '/home/milottit/Libraries';
        logger.log((await lookForLocalPackages(libraryFolder, {
            excludeFolder: [ '.git', '.vscode', 'node_modules', ],
            filterPackage: project => project.package.name.startsWith('@upradata/')
        })).map(p => p.folder));
        process.exit(1);
    })(); */
}
