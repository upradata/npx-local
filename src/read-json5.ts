import JSON5 from 'json5';
import { readFileSync, readFile } from 'fs-extra';
import { JSONSchemaForNPMPackageJsonFiles } from '@schemastore/package';


export function readJson(filename: string, mode: 'sync' | 'async'): Promise<JSONSchemaForNPMPackageJsonFiles> | JSONSchemaForNPMPackageJsonFiles {

    const normalizeData = data => data === '' ? {} : JSON5.parse(data);

    if (mode === 'sync') {
        const data = readFileSync(filename, 'utf8');
        return normalizeData(data);
    }

    return readFile(filename, 'utf8').then(normalizeData);
}

export function readJsonSync(filename: string): JSONSchemaForNPMPackageJsonFiles {
    return readJson(filename, 'sync') as JSONSchemaForNPMPackageJsonFiles;
}

export function readJsonASync(filename: string): Promise<JSONSchemaForNPMPackageJsonFiles> {
    return readJson(filename, 'async') as Promise<JSONSchemaForNPMPackageJsonFiles>;
}

/* export async function readJson(directory: string, filename: string): Promise<JSONSchemaForNPMPackageJsonFiles> {
    const file = path.join(directory, filename);

    try {
        if (!await pathExists(file)) {
            const msg = `${file} doesn't exist`;
            console.error(red`${msg}`);
            throw new Error(msg);
        }

        const data = await readFile$(file, 'utf8');
        return data === '' ? {} : JSON5.parse(data);
    } catch (e) {
        throw e;
    }
}
 */
