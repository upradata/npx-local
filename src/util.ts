import { isAbsolute, relative, resolve } from 'path';
import { DependencyName, DependencyType, RelativeAbsolute, SemVersionType } from './types';


const dependencyDef = <D extends { depType: DependencyType; flags: string; depName: DependencyName; }>(def: D) => def;

export const dependenciesDef = [
    dependencyDef({ depType: 'prod', flags: '-P, --prod', depName: 'dependencies' } as const),
    dependencyDef({ depType: 'dev', flags: '-D, --dev', depName: 'devDependencies' } as const),
    dependencyDef({ depType: 'peers', flags: '--peers', depName: 'peersDependencies' } as const),
] as const;

export const getDependencyName = (dependencyType: DependencyType) => dependenciesDef.find(d => d.depType === dependencyType)?.depName;



const semVerDefinition = <D extends { name: SemVersionType; flags: string; }>(def: D) => def;

export const semVerDefinitions = [
    semVerDefinition({ name: 'major', flags: '-M, --major' } as const),
    semVerDefinition({ name: 'minor', flags: '-m, --minor' } as const),
    semVerDefinition({ name: 'major', flags: '-p, --patch', } as const),
] as const;



export const relativeAbsolutPath = (path: string, dir: string = process.cwd()): RelativeAbsolute => {
    return {
        relative: isAbsolute(path) ? relative(dir, path) : path,
        absolute: isAbsolute(path) ? path : resolve(dir, path)
    };
};
