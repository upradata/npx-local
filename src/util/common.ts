export const terminalWidth = process.stdout.columns - 1 || 80;

export function stringAlignCenter(s: string, size: number = terminalWidth): string {

    const trim = s.trim();
    const whitespaceWidth = size - trim.length;

    if (whitespaceWidth <= 0)
        return s;

    const beginL = Math.floor(whitespaceWidth / 2) - 1; // -1 je ne sais pas pk :)
    const endL = whitespaceWidth - beginL;


    return ' '.repeat(beginL) + trim + ' '.repeat(endL);
}
