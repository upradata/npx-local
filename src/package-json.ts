import { join } from 'path';
// eslint-disable-next-line import/no-extraneous-dependencies
import { JSONSchemaForNPMPackageJsonFiles } from '@schemastore/package';
import {
    fileExists,
    findUp,
    readPackageJson,
    SyncAsync,
    SyncAsyncMode,
    SyncAsyncType
} from '@upradata/node-util';
import { CodifiedError, ObjectOf, ValueOf } from '@upradata/util';
import { Dependency } from './local-dependency';
import { getOptions } from './local-install.options';
import { Errors } from './types';


export class Local {
    dependencies?: ObjectOf<Dependency>;
    usedBy?: ObjectOf<string>;
}


export type LocalInstallPackageJsonType = JSONSchemaForNPMPackageJsonFiles & { local?: Local; };


export class LocalInstallPackageJson {
    public json: LocalInstallPackageJsonType;
    public _path: string;


    constructor(public directory: string) { }

    public async load() {
        return this.readJson();
    }


    private getPackageJsonPath<M extends SyncAsyncMode>(mode: M) {

        const path = <Mode extends SyncAsyncMode>(m: Mode): SyncAsyncType<Mode, string> => {
            const pkgJsonFile = 'package.json';

            if (getOptions().findUp)
                return findUp[ m ](pkgJsonFile, { type: 'file', cwd: this.directory }) as any;

            const pkgJsonPath = join(this.directory, pkgJsonFile);

            return (m === 'sync' ? pkgJsonPath : Promise.resolve(pkgJsonPath)) as any;
        };

        const set = (p: string) => { this._path = p; return p; };

        const r = mode === 'sync' ?
            this._path ? this._path : set(path('sync')) :
            this._path ? Promise.resolve(this._path) : path('async').then(set);

        return r as SyncAsyncType<M, string>;
    }

    public get path(): SyncAsync<string> {
        const _this = this;

        return {
            get sync() { return _this.getPackageJsonPath('sync'); },
            get async() { return _this.getPackageJsonPath('async'); }
        };
    }


    async readJson(force: boolean = false) {
        if (!this.json || force) {
            const path = await this.path.async;

            if (await fileExists.async(path)) {

                const json = await readPackageJson.async(await this.path.async);
                this.json = json as LocalInstallPackageJson;

            } else {

                throw new CodifiedError<Errors>({
                    message: `The directory "${this.directory}" does not have a "package.json"`,
                    code: Errors.NO_PACKAGE_JSON
                });
            }
        }

        return this.json;
    }

    private getProp = <O, P extends keyof O>(o: O, prop: P): O[ P ] => {
        o[ prop ] = o[ prop ] || {} as any;
        return o[ prop ];
    };

    private getLocal<M extends SyncAsyncMode>(mode: M) {
        const get = (json: LocalInstallPackageJsonType) => this.getProp(json, 'local');

        const r = mode === 'sync' ? get(this.json) : this.readJson().then(get);
        return r as SyncAsyncType<M, Local>;
    }

    private getLocalProp<M extends SyncAsyncMode>(mode: M, prop: keyof Local) {
        const get = (local: Local) => this.getProp(local, prop);

        const r = mode === 'sync' ? get(this.getLocal('sync')) : this.getLocal('async').then(get);
        return r as SyncAsyncType<M, ValueOf<Local>>;
    }

    public get local(): SyncAsync<Local> {
        const _this = this;

        return {
            get sync() { return _this.getLocal('sync'); },
            get async() { return _this.getLocal('async'); }
        };
    }

    public localProp(prop: keyof Local): SyncAsync<ValueOf<Local>> {
        const _this = this;

        return {
            get sync() { return _this.getLocalProp('sync', prop); },
            get async() { return _this.getLocalProp('async', prop); }
        };
    }

}
