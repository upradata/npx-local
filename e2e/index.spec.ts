import { projectDir } from './util';
import { readJsonAsync } from '../src/read-json5';
import path from 'path';
import { LocalInstallPackageJsonType } from '../src/package-json';
import { chain } from '@upradata/util';
import { execNpmLocal, checkNodeModules, Dependency, checkLocalProp } from './common';
import { projects } from './projects.config';
import { remove } from 'fs-extra';


// execSync(`cd ${root} && tsc`);

async function checkLocalDependencies(projectI: number, deps: Dependency[]) {
    const packageJson = await readJsonAsync(path.join(projectDir(projectI), 'package.json')) as LocalInstallPackageJsonType;
    // console.log({ projectI, deps });
    checkLocalProp(chain(() => packageJson.local.dependencies), deps.map(dep => dep.projectI));
    await checkNodeModules(projectI, deps);
}


async function checkLocalUsedBy(projectI: number, usedBys: number[]) {
    const packageJson = await readJsonAsync(path.join(projectDir(projectI), 'package.json')) as LocalInstallPackageJsonType;
    checkLocalProp(chain(() => packageJson.local.usedBy), usedBys);
}

async function snapshot(projectI: number) {
    const packageJson = await readJsonAsync(path.join(projectDir(projectI), 'package.json')) as LocalInstallPackageJsonType;
    expect(packageJson).toMatchSnapshot();
}


describe('Test => npmlocal --verbose --force [...local-projects]', () => {

    beforeAll(async () => {
        jest.setTimeout(30000);
        await Promise.all([ 1, 2, 3, 4 ].map(i => remove(path.join(projectDir(i), 'node_modules')).catch(console.warn)));
        await execNpmLocal(1, [ 2, 3, 4 ]);
        await execNpmLocal(4, [ 1, 2, 3 ]);
    });

    test('is localDependencies set', async () => {
        await Promise.all([
            checkLocalDependencies(projects.project1.projectI, projects.project1.localDeps),
            checkLocalDependencies(projects.project4.projectI, projects.project4.localDeps)
        ]);
    });


    test('is usedBy set', async () => {
        await Promise.all([
            checkLocalUsedBy(projects.project1.projectI, projects.project1.usedBys),
            checkLocalUsedBy(projects.project2.projectI, projects.project2.usedBys),
            checkLocalUsedBy(projects.project3.projectI, projects.project3.usedBys),
            checkLocalUsedBy(projects.project4.projectI, projects.project4.usedBys),
        ]);
    });

    test('is package.json same snapshot', async () => {
        await Promise.all(Object.values(projects).map((p, i) => snapshot(i + 1)));
    });
}
);

describe('Test => npmlocal in current npm package', () => {

    beforeAll(async () => {
        jest.setTimeout(30000);
        await remove(path.join(projectDir(4), 'node_modules')).catch(console.warn);
        await execNpmLocal(4);
    });

    test('Install package.json', async () => {
        await checkLocalDependencies(projects.project4.projectI, projects.project4.localDeps);
    });
}
);
