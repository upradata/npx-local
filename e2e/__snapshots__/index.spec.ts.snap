// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test => npmlocal --verbose --force [...local-projects] is package.json same snapshot 1`] = `
Object {
  "author": "Thomas Milotti <thomas.milotti@gmail.com> (https://github.com/milottit)",
  "description": "",
  "license": "MIT",
  "local": Object {
    "dependencies": Object {
      "project2": "link:/home/milottit/Libraries/LocalInstallation/TestProjects/Project2@0.0.1",
      "project3": "link:/home/milottit/Libraries/LocalInstallation/TestProjects/Project3@0.0.1",
      "project4": "link:/home/milottit/Libraries/LocalInstallation/TestProjects/Project4@0.0.1",
    },
    "usedBy": Object {
      "project4": "/home/milottit/Libraries/LocalInstallation/TestProjects/Project4",
    },
  },
  "main": "lib",
  "name": "project1",
  "scripts": Object {
    "test": "echo \\"Error: no test specified\\" && exit 1",
  },
  "version": "0.0.1",
}
`;

exports[`Test => npmlocal --verbose --force [...local-projects] is package.json same snapshot 2`] = `
Object {
  "author": "Thomas Milotti <thomas.milotti@gmail.com> (https://github.com/milottit)",
  "description": "",
  "license": "MIT",
  "local": Object {
    "usedBy": Object {
      "project1": "/home/milottit/Libraries/LocalInstallation/TestProjects/Project1",
      "project4": "/home/milottit/Libraries/LocalInstallation/TestProjects/Project4",
    },
  },
  "module": "lib-esm/index.js",
  "name": "project2",
  "scripts": Object {
    "test": "echo \\"Error: no test specified\\" && exit 1",
  },
  "version": "0.0.1",
}
`;

exports[`Test => npmlocal --verbose --force [...local-projects] is package.json same snapshot 3`] = `
Object {
  "author": "Thomas Milotti <thomas.milotti@gmail.com> (https://github.com/milottit)",
  "description": "",
  "files": Array [
    "lib",
    "lib-esm",
    "README.md",
  ],
  "license": "MIT",
  "local": Object {
    "usedBy": Object {
      "project1": "/home/milottit/Libraries/LocalInstallation/TestProjects/Project1",
      "project4": "/home/milottit/Libraries/LocalInstallation/TestProjects/Project4",
    },
  },
  "main": "lib",
  "module": "lib-esm",
  "name": "project3",
  "scripts": Object {
    "test": "echo \\"Error: no test specified\\" && exit 1",
  },
  "version": "0.0.1",
}
`;

exports[`Test => npmlocal --verbose --force [...local-projects] is package.json same snapshot 4`] = `
Object {
  "author": "Thomas Milotti <thomas.milotti@gmail.com> (https://github.com/milottit)",
  "dependencies": Object {
    "node-watch": "^0.6.3",
  },
  "description": "",
  "license": "MIT",
  "local": Object {
    "dependencies": Object {
      "project1": "link:/home/milottit/Libraries/LocalInstallation/TestProjects/Project1@0.0.1",
      "project2": "link:/home/milottit/Libraries/LocalInstallation/TestProjects/Project2@0.0.1",
      "project3": "link:/home/milottit/Libraries/LocalInstallation/TestProjects/Project3@0.0.1",
    },
    "usedBy": Object {
      "project1": "/home/milottit/Libraries/LocalInstallation/TestProjects/Project1",
    },
  },
  "main": "lib/index.js",
  "name": "project4",
  "scripts": Object {
    "test": "echo \\"Error: no test specified\\" && exit 1",
  },
  "version": "0.0.1",
}
`;
